src/main/java/it/gov/pagopa/ranking/service/RankingMaterializerServiceImpl.java

```diff
@@ -0,0 +1,117 @@
+
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.List;
+
+@Service
+@Slf4j
+public class RankingMaterializerServiceImpl implements RankingMaterializerService {
+
+    private final OnboardingRankingRequestsRepository onboardingRankingRequestsRepository;
+    private final int size;
+    private final OnboardingRankingRequests2RankingCsvDTOMapper csvMapper;
+    private final RankingCsvWriterService csvWriterService;
+
+    public RankingMaterializerServiceImpl(
+            OnboardingRankingRequestsRepository onboardingRankingRequestsRepository,
+            @Value("${app.ranking.query-page-size}")  int size,
+            OnboardingRankingRequests2RankingCsvDTOMapper csvMapper,
+            RankingCsvWriterService csvWriterService
+    ) {
+        this.onboardingRankingRequestsRepository = onboardingRankingRequestsRepository;
+        this.size = size;
+        this.csvMapper = csvMapper;
+        this.csvWriterService = csvWriterService;
+    }
+
+    @Override
+    public String materialize(InitiativeConfig initiativeConfig) {
+
+        try (FileWriter outputCsvWriter = new FileWriter("tmp")) { // TODO define fileName
+            String initiativeId = initiativeConfig.getInitiativeId();
+            long totalEligibleOk = initiativeConfig.getTotalEligibleOk();
+            long totalEligibleKo = initiativeConfig.getTotalEligibleKo();
+
+            int page = 0;
+            int rank = 1;
+            List<OnboardingRankingRequests> pageContent;
+            while (!(pageContent = onboardingRankingRequestsRepository.findAllByInitiativeId(
+                    initiativeId,
+                    PageRequest.of(page++, size, getSorting(initiativeConfig))
+            )).isEmpty()) {
+
+                log.info("[RANKING_MATERIALIZER] Reading page number {} of initiative with id {}", page, initiativeId);
+                for (OnboardingRankingRequests r : pageContent) {
+                    int actualRank = rank++;
+
+                    // updating the entity only if changed
+                    if (r.getRank() != actualRank) {
+                        r.setRank(actualRank);
+
+                        if (r.getBeneficiaryRankingStatus() != BeneficiaryRankingStatus.ONBOARDING_KO) {
+                            if (r.getRank() <= initiativeConfig.getSize()) {
+                                r.setBeneficiaryRankingStatus(BeneficiaryRankingStatus.ELIGIBLE_OK);
+                                ++totalEligibleOk;
+                            } else {
+                                r.setBeneficiaryRankingStatus(BeneficiaryRankingStatus.ELIGIBLE_KO);
+                                ++totalEligibleKo;
+                            }
+                        }
+                    }
+                }
+                onboardingRankingRequestsRepository.saveAll(pageContent);
+
+                csvWriterService.write(buildCsvLines(pageContent), outputCsvWriter); // TODO this method should write directly into the file
```

Comment:
let's store in the db and write in the csv at the same moment inside the "better solution" described above, in order to not perform all of these actions once with the entire page